<a href="https://colab.research.google.com/github/shahabday/DSR-time-series/blob/main/MAPE_and_SMAPE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Metrics for Time Series Forecasting - MAPE and SMAPE

### What is MAPE?

Mean Absolute Percentage Error (MAPE) is a metric used to evaluate the accuracy of a forecasting model. MAPE provides a percentage-based scale to measure the error between actual and forecasted values. However, MAPE is not symmetric, meaning it can be biased towards underestimates or overestimates.

The formula for calculating MAPE is:

$$
\text{MAPE} = \frac{100}{N} \sum_{t=1}^{N} \frac{\left| A_t - F_t \right|}{\left| A_t \right|}
$$

Here:
- $ A_t $ is the actual value at time $ t $
- $ F_t $ is the forecasted value at time $ t $
- $ N $ is the number of data points

### Examples

Let's consider the same two example scenarios:

1. **Scenario 1**: The actual values are $[100, 200, 300, 400]$ and the forecasted values are $[110, 190, 290, 410]$.
2. **Scenario 2**: The actual values are $[10, 20, 30, 40]$ and the forecasted values are $[11, 19, 29, 41]$.

#### Scenario 1 Calculation

For the first data point:

$$
\frac{\left| 100 - 110 \right|}{\left| 100 \right|} = \frac{10}{100} = 0.1
$$

Similarly, we calculate for other data points and then take the mean.

#### Scenario 2 Calculation

For the first data point:

$$
\frac{\left| 10 - 11 \right|}{\left| 10 \right|} = \frac{1}{10} = 0.1
$$

Again, we calculate for other data points and then take the mean.

### Code Example

Here is a Python code snippet to calculate MAPE:

In [ ]:
import numpy as np

def mape(actual, forecast):
    # Ensure the arrays have the same length
    assert len(actual) == len(forecast), "Length of actual and forecast arrays should be the same"

    # Avoid division by zero by replacing zeros in the 'actual' array with a small number
    actual = np.where(actual == 0, 1e-10, actual)

    # Calculate MAPE for each data point
    mape_values = np.abs((actual - forecast) / actual)

    # Calculate the mean MAPE value
    mean_mape = np.mean(mape_values) * 100  # Multiply by 100 to get percentage

    return mean_mape

# Scenario 1
actual_1 = np.array([100, 200, 300, 400])
forecast_1 = np.array([110, 190, 290, 410])
mape_1 = mape(actual_1, forecast_1)

# Scenario 2
actual_2 = np.array([10, 20, 30, 40])
forecast_2 = np.array([11, 19, 29, 41])
mape_2 = mape(actual_2, forecast_2)

mape_1, mape_2


(5.208333333333334, 5.208333333333334)

Now, try replacing an actual value with a zero. What happens to MAPE?

### What is SMAPE?

Symmetric Mean Absolute Percentage Error (SMAPE) is a metric used to evaluate the accuracy of a forecasting method. Unlike Mean Absolute Error (MAE) or Mean Squared Error (MSE), SMAPE takes into account the relative error between the predicted and the actual values. SMAPE is symmetric, meaning it treats over-forecasting and under-forecasting equally. This is in contrast to MAPE (Mean Absolute Percentage Error), which is not symmetric and can be biased toward underestimates.

The formula for calculating SMAPE is:

$$
\text{SMAPE} = \frac{100}{N} \sum_{t=1}^{N} \frac{\left| A_t - F_t \right|}{\frac{\left| A_t \right| + \left| F_t \right|}{2}}
$$

Here:
- $ A_t $ is the actual value at time $ t $
- $ F_t $ is the forecasted value at time $ t $
- $ N $ is the number of data points

### Examples

Let's consider two example scenarios:

1. **Scenario 1**: The actual values are $[100, 200, 300, 400]$ and the forecasted values are $[110, 190, 290, 410]$.
2. **Scenario 2**: The actual values are $[10, 20, 30, 40]$ and the forecasted values are $[11, 19, 29, 41]$.

#### Scenario 1 Calculation

For the first data point:

$$
\frac{\left| 100 - 110 \right|}{\frac{\left| 100 \right| + \left| 110 \right|}{2}} = \frac{10}{105} \approx 0.0952
$$

Similarly, we calculate for other data points and then take the mean.

#### Scenario 2 Calculation

For the first data point:

$$
\frac{\left| 10 - 11 \right|}{\frac{\left| 10 \right| + \left| 11 \right|}{2}} = \frac{1}{10.5} \approx 0.0952
$$

Again, we calculate for other data points and then take the mean.

### Code Example

Here is a Python code snippet to calculate SMAPE:



In [ ]:
import numpy as np

def smape(actual, forecast):
    # Ensure the arrays have the same length
    assert len(actual) == len(forecast), "Length of actual and forecast arrays should be the same"

    # Calculate SMAPE for each data point
    smape_values = 2 * np.abs(actual - forecast) / (np.abs(actual) + np.abs(forecast))

    # Calculate the mean SMAPE value
    mean_smape = np.mean(smape_values) * 100  # Multiply by 100 to get percentage

    return mean_smape

# Scenario 1
actual_1 = np.array([100, 200, 300, 0])
forecast_1 = np.array([110, 190, 290, 410])
smape_1 = smape(actual_1, forecast_1)

# Scenario 2
actual_2 = np.array([10, 20, 30, 40])
forecast_2 = np.array([11, 19, 29, 41])
smape_2 = smape(actual_2, forecast_2)

smape_1, smape_2

(54.5104612901223, 5.12774524073959)

Once again, try replacing an actual value with a zero. What happens to SMAPE?

### Exercise: Compute MAPE and SMAPE on a Dataset and Compare

#### Objective:

Your task is to compute both the MAPE (Mean Absolute Percentage Error) and SMAPE (Symmetric Mean Absolute Percentage Error) on a given dataset. After calculating these metrics, you should provide a comparison and discuss the advantages and disadvantages of each in the context of the dataset.

#### Dataset:

Consider the following time series data for actual sales and forecasted sales (in units) for a retail store:

| Week  | Actual Sales | Forecasted Sales |
|-------|-------------|------------------|
| 1     | 120         | 130              |
| 2     | 135         | 128              |
| 3     | 140         | 142              |
| 4     | 130         | 127              |
| 5     | 125         | 123              |

In [ ]:
# Given dataset
actual_sales = np.array([120, 135, 140, 130, 125])
forecasted_sales = np.array([130, 128, 142, 127, 123])

# Compute MAPE and SMAPE
mape_value = mape(actual_sales, forecasted_sales)
smape_value = smape(actual_sales, forecasted_sales)

# Output the computed values
print(f"MAPE: {mape_value}")
print(f"SMAPE: {smape_value}")

MAPE: 3.770956450956451
SMAPE: 3.7378334417325743


To get a better sense of their difference, experiment playing with both actual and forecast values, and observe the corresponding impact on the error metrics.

### Solution to the Exercise

#### Computed MAPE and SMAPE Values:

- MAPE: \(3.77\%\)
- SMAPE: \(3.74\%\)

#### Comparison and Discussion:

1. **Similarity in Values**: Both MAPE and SMAPE values are quite similar for this dataset. This indicates that the forecast errors are fairly symmetric around the actual values.

2. **Scale Sensitivity**: MAPE can be sensitive to the scale of the actual values. In scenarios where the actual values are closer to zero, MAPE can be highly skewed. SMAPE is relatively less sensitive to the scale of the actual values.

3. **Symmetry**: SMAPE treats over-forecasting and under-forecasting symmetrically, which is an advantage if both types of errors are equally important for your application. MAPE can be biased and is not symmetric.

4. **Interpretability**: Both metrics are easy to interpret as they provide percentage errors. However, MAPE can become undefined or extremely large when the actual values approach zero, whereas SMAPE avoids this issue.

In summary, the choice between MAPE and SMAPE would depend on the specific requirements of your forecasting task. If you want a symmetric measure and are concerned about the scale of your actual values, SMAPE could be a better choice. On the other hand, if you are more concerned with absolute percentage errors, then MAPE would be more appropriate.